code/modules/reagents/reagent_containers.dm

```diff
@@ -47,17 +53,35 @@
 		reagents.add_reagent_list(list_reagents)
 
 /obj/item/reagent_containers/attack_self(mob/user)
-	if(possible_transfer_amounts.len)
-		var/i=0
-		for(var/A in possible_transfer_amounts)
-			i++
-			if(A == amount_per_transfer_from_this)
-				if(i<possible_transfer_amounts.len)
-					amount_per_transfer_from_this = possible_transfer_amounts[i+1]
-				else
-					amount_per_transfer_from_this = possible_transfer_amounts[1]
-				balloon_alert(user, "Transferring [amount_per_transfer_from_this]u")
-				return
+	change_transfer_amount(user, FORWARD)
+
+/obj/item/reagent_containers/attack_self_secondary(mob/user)
+	change_transfer_amount(user, BACKWARD)
+
+/obj/item/reagent_containers/proc/mode_change_message(mob/user)
+	return
+
+/obj/item/reagent_containers/proc/change_transfer_amount(mob/user, direction)
+	if(!possible_transfer_amounts.len)
+		return
+	var/list_position = 0
+	for(var/transfer_amount in possible_transfer_amounts)
+		list_position++
+		if(transfer_amount == amount_per_transfer_from_this)
+			switch(direction)
+				if(FORWARD)
+					if(list_position < possible_transfer_amounts.len) //not at end
+						amount_per_transfer_from_this = possible_transfer_amounts[list_position + 1] //move to next entry
+					else
+						amount_per_transfer_from_this = possible_transfer_amounts[1] //move to list start
+				if(BACKWARD)
+					if(list_position == 1) //at list start
+						amount_per_transfer_from_this = possible_transfer_amounts[possible_transfer_amounts.len] //move to list end
+					else
+						amount_per_transfer_from_this = possible_transfer_amounts[list_position - 1] //move to previous entry
+			balloon_alert(user, "transferring [amount_per_transfer_from_this]u")
+			mode_change_message(user)
+			return
```

Comment:
I'd keep the current list_position as an object variable, personally, instead of searching for the position each time.
Tne you can simply do this:


```suggestion
/obj/item/reagent_containers/proc/change_transfer_amount(mob/user, direction)
	var/list_len = length(possible_transfer_amounts)
	if(!list_len)
		return
	switch(direction)
		if(FORWARD)
			list_position = (list_position % list_len) + 1
		if(BACKWARD)
			list_position = (list_position - 1) || list_len
	amount_per_transfer_from_this = possible_transfer_amounts[list_position]
	balloon_alert(user, "transferring [amount_per_transfer_from_this]u")
	mode_change_message(user)
```
The modulo operator will result in `0` if the `list_position` value matches the length, and thus increment it back to `1`, looping from last to first position.
The `||` operator will substitute the value if falsey (as `0`) for the one on the right side, this being the length of the list, thus looping from first to last position.

I'd also turn `possible_transfer_amounts` into a lazy list as well, but that may be out of scope.

Reply:
Thank you, this is clever